From d03dc60a7a5b2794357cd6db187023163d035b32 Mon Sep 17 00:00:00 2001
From: Marcelo Leitner <mleitner@redhat.com>
Date: Tue, 21 Oct 2014 16:20:50 -0400
Subject: [net] netfilter: nf_tables: store and dump set policy

Message-id: <20141021162050.22888.53483.stgit@localhost.localdomain>
Patchwork-id: 98007
O-Subject: [RHEL7 net PATCH 110/124] netfilter: nf_tables: store and dump set policy
Bugzilla: 1148041 1155088
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

Backport of upstream commit: 9363dc4b599949bde338cdaba1cf7cac243e4e97

 commit 9363dc4b599949bde338cdaba1cf7cac243e4e97
 Author: Arturo Borrero <arturo.borrero.glez@gmail.com>
 Date:   Tue Sep 23 13:30:41 2014 +0200

    netfilter: nf_tables: store and dump set policy

    We want to know in which cases the user explicitly sets the policy
    options. In that case, we also want to dump back the info.

    Signed-off-by: Arturo Borrero Gonzalez <arturo.borrero.glez@gmail.com>
    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Marcelo Ricardo Leitner <mleitner@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/include/net/netfilter/nf_tables.h b/include/net/netfilter/nf_tables.h
index c4d8619..3d72923 100644
--- a/include/net/netfilter/nf_tables.h
+++ b/include/net/netfilter/nf_tables.h
@@ -241,6 +241,7 @@ void nft_unregister_set(struct nft_set_ops *ops);
  *  @dtype: data type (verdict or numeric type defined by userspace)
  *  @size: maximum set size
  *  @nelems: number of elements
+ * @policy: set parameterization (see enum nft_set_policies)
  *  @ops: set ops
  *  @flags: set flags
  *  @klen: key length
@@ -255,6 +256,7 @@ struct nft_set {
  u32    dtype;
  u32    size;
  u32    nelems;
+ u16    policy;
  /* runtime data below here */
  const struct nft_set_ops *ops ____cacheline_aligned;
  u16    flags;
diff --git a/net/netfilter/nf_tables_api.c b/net/netfilter/nf_tables_api.c
index 7737122..c0831f6 100644
--- a/net/netfilter/nf_tables_api.c
+++ b/net/netfilter/nf_tables_api.c
@@ -2344,6 +2344,11 @@ static int nf_tables_fill_set(struct sk_buff *skb, const struct nft_ctx *ctx,
    goto nla_put_failure;
  }
 
+ if (set->policy != NFT_SET_POL_PERFORMANCE) {
+  if (nla_put_be32(skb, NFTA_SET_POLICY, htonl(set->policy)))
+   goto nla_put_failure;
+ }
+
  desc = nla_nest_start(skb, NFTA_SET_DESC);
  if (desc == NULL)
   goto nla_put_failure;
@@ -2669,6 +2674,7 @@ static int nf_tables_newset(struct sock *nlsk, struct sk_buff *skb,
  set->dlen  = desc.dlen;
  set->flags = flags;
  set->size  = desc.size;
+ set->policy = policy;
 
  err = ops->init(set, &desc, nla);
  if (err < 0)
-- 
1.7.1